src/funnels/Swapper.sol

```diff
@@ -0,0 +1,243 @@
+    uint256 public maxOut;     // [WAD]       Max amount swapped from gem to nst every hop
+    uint256 public wantIn;     // [WAD]       Relative multiplier of the reference price (equal to 1 gem/nst) to insist on in the swap from nst to gem.
+    uint256 public wantOut;    // [WAD]       Relative multiplier of the reference price (equal to 1 nst/gem) to insist on in the swap from gem to nst.
+    uint256 public zzz;        // [Timestamp] Last swap
+    uint256 public fee;        // [BPS]       UniV3 pool fee
+
+    uint256[] internal weights;   // Bit vector tightly packing (address(box), uint96(percentage)) tuple entries such that the sum of all percentages = 1 WAD (100%)
+
+    event Rely  (address indexed usr);
+    event Deny  (address indexed usr);
+    event Kissed(address indexed usr);
+    event Dissed(address indexed usr);
+    event Permit(address indexed usr);
+    event Forbid(address indexed usr);
+    event File  (bytes32 indexed what, uint256 data);
+    event File  (bytes32 indexed what, address data);
+    event File  (bytes32 indexed what, address data, uint256 val);
+    event Swap  (address indexed kpr, address indexed from, address indexed to, uint256 wad, uint256 out);
+    event Quit  (address indexed usr, uint256 wad);
+
+    struct Weight {
+        address box;
+        uint96 wad; // percentage in WAD such that 1 WAD = 100%
+    }
+
+    constructor(address _nst, address _gem, address _uniV3Router) {
+        nst = _nst;
+        gem = _gem;
+        uniV3Router = _uniV3Router;
+
+        GemLike(nst).approve(uniV3Router, type(uint256).max);
+        GemLike(gem).approve(uniV3Router, type(uint256).max);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "Swapper/not-authorized");
+        _;
+    }
+
+    // permissionned to whitelisted facilitators
+    modifier toll { 
+        require(buds[msg.sender] == 1, "Swapper/non-facilitator"); 
+        _;
+    }
+
+    // permissionned to whitelisted keepers
+    modifier keeper { 
+        require(keepers[msg.sender] == 1, "Swapper/non-keeper"); 
+        _;
+    }
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    address public immutable uniV3Router;
+    address public immutable nst;
+    address public immutable gem;
+
+    function rely(address usr)   external auth { wards[usr]   = 1; emit Rely(usr); }
+    function deny(address usr)   external auth { wards[usr]   = 0; emit Deny(usr); }
+    function kiss(address usr)   external auth { buds[usr]    = 1; emit Kissed(usr); }
+    function diss(address usr)   external auth { buds[usr]    = 0; emit Dissed(usr); }
+    function permit(address usr) external toll { keepers[usr] = 1; emit Permit(usr); }
+    function forbid(address usr) external toll { keepers[usr] = 0; emit Forbid(usr); }
+
+    function file(bytes32 what, uint256 data) external auth {
```

Comment:
Maybe we should consider new parameters for the facilitator to set - max amount of swaps to be done in each direction. So something like 2 countdown variables named `countIn` and `countOut`.

That will allow doing just one swap if needed, and also better control of how much funds to move in general.
Also, currently it's pretty weird that the weights are used to both direction but the facilitator doesn't define which is the allowed direction. That will solve that as well as probably one of the directions will have a count of 0 set up.

Reply:
Yes, make sense so as to reduce the impact of a compromised keeper.

Reply:
Done